platform/ext/target/nordic_nrf/common/core/tfm_hal_its_encryption.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "platform/include/tfm_platform_system.h"
+#include "platform/include/tfm_hal_its.h"
+#include "psa/crypto.h"
+
+#include "nrf_cc3xx_platform_derived_key.h"
+#include "nrf_cc3xx_platform.h"
+#include "nrf_cc3xx_platform_kmu.h"
+#include <string.h>
+
+#define ITS_ENCRYPTION_SUCCESS 0
+
+#define HUK_KMU_SLOT 2
+#define HUK_KMU_SIZE_BITS 128
+
+/* Global encryption counter which resets per boot. The counter ensures that
+ * the nonce will not be identical for consecutive file writes during the same
+ * boot.
+ */
+static uint32_t g_enc_counter;
+
+/* The global nonce seed which is fetched once in every boot. The seed is used
+ * as part of the nonce and allows the platforms to diversify their nonces
+ * across resets. Note that the way that this seed is generated is platform
+ * specific, so the diversification is optional.
+ */
+static uint8_t g_enc_nonce_seed[TFM_ITS_ENC_NONCE_LENGTH -
+                                sizeof(g_enc_counter)];
+
+
+enum tfm_hal_status_t tfm_hal_its_aead_generate_nonce(uint8_t *nonce,
+                                                      size_t nonce_size)
+{
+    int err;
+
+    if(nonce == NULL){
+        return TFM_HAL_ERROR_INVALID_INPUT;
+    }
+
+    if(nonce_size < sizeof(g_enc_nonce_seed) + sizeof(g_enc_counter)){
+        return TFM_HAL_ERROR_INVALID_INPUT;
+    }
+
+    if (g_enc_counter == 0) {
+        if (sizeof(g_enc_nonce_seed) != 8) {
+            return TFM_HAL_ERROR_GENERIC;
+        }
+
+        err = nrf_cc3xx_platform_get_nonce_seed(g_enc_nonce_seed);
+        if (err != 0) {
+            return TFM_HAL_ERROR_GENERIC;
+        }
+    }
+
+    memcpy(nonce, g_enc_nonce_seed, sizeof(g_enc_nonce_seed));
+    memcpy(nonce + sizeof(g_enc_nonce_seed),
+               &g_enc_counter,
+               sizeof(g_enc_counter));
+
+    g_enc_counter++;
```

Comment:
The docs for tfm_hal_its_aead_generate_nonce highly recommend a PRNG.

So shouldn't we use that instead of
RNG + 0, RNG + 1, RNG + 2

i++

isn't a very good PRNG IMHO.

Reply:
The main goal here was that we wanted to have unique IV for each file across resets. The fact that the IV is just IV++ is not an issue since it is still unique per reset and per file. The reason that we didn't use the RPNG for every call here was that is more costly and we didn't consider it adding security to the scheme. By using a unique PRNG per boot makes it nearly impossible to have the same IV for the same file twice across boots and that's enough. 

We can rephrase the documentation text to say that we highly suggest having a unique IV per file which is not the same across resets.

Reply:
Thanks for clarifying,  I rephrased the docs, but I like the hint with the PRN.